Libraries/LibGUI/TextBox.cpp

```diff
@@ -37,4 +37,67 @@ TextBox::~TextBox()
 {
 }
 
+void TextBox::keydown_event(GUI::KeyEvent& event)
+{
+    TextEditor::keydown_event(event);
+
+    if (event.key() == Key_Up) {
+        if (on_up_pressed)
+            on_up_pressed();
+
+        if (!m_history_enabled || !has_any_history() || !can_go_backwards_in_history())
+            return;
+
+        if (m_history_index >= static_cast<int>(m_input_history.size()))
+            m_saved_input = text();
+
+        m_history_index--;
+        set_text(current_historical_input());
+    } else if (event.key() == Key_Down) {
+        if (on_down_pressed)
+            on_down_pressed();
+
+        if (!m_history_enabled || !has_any_history())
+            return;
+
+        if (can_go_forwards_in_history()) {
+            m_history_index++;
+            set_text(current_historical_input());
+        } else if (has_current_historical_input()) {
+            m_history_index++;
+            set_text(m_saved_input);
+        }
+    }
+}
+
+void TextBox::add_current_text_to_history()
+{
+    if (!m_history_enabled)
+        return;
+    
+    auto input = text();
+    if (!has_any_history() || m_input_history[m_input_history.size() - 1] != input)
```

Comment:
This is also a bit muddy. "if not has any history" is a bit hard for my brain to parse. :)
Suggestion:
```cpp
if (m_input_history.is_empty() || m_input_history.last() != input)
```

Reply:
Well, it seems we're inverting the return value of `has_any_history` everywhere it's used, would `if (has_no_history() || m_input_history.last() != input)` be reasonable, or should I just use the `is_empty` method on the history list?

Reply:
Let's just use the history vector directly. We could even rename it to `m_history` since there is only one kind of history here anyway :)